### ISSUE TITLE
Fix `stylelint.utils.checkAgainstRule()` for Promise-based rules

### ISSUE THREAD

**Issue by aaronccasanova** [stylelint/stylelint#7820]
### What minimal example or steps are needed to reproduce the bug?

Call `checkAgainstRule` with an async custom rule and notice no warnings are reported.

In this pseudo code example `custom/rule-one` won't report warnings from `custom/rule-two`:

```js
export const ruleName = 'custom/rule-one';

export const {rule} = stylelint.createPlugin(
  ruleName,
  (ruleSettings) => async (root, result) => {
    await stylelint.utils.checkAgainstRule(
      {
        ruleName: 'custom/rule-two',
        ruleSettings,
        root,
      },
      (warning) => {
        stylelint.utils.report({
          ruleName,
          result,
          node: warning.node,
          message: warning.text,
        });
      },
    );
  };
);
```
```js
export const ruleName = 'custom/rule-two';

export const {rule} = stylelint.createPlugin(
  ruleName,
  (ruleSettings) => async (root, result) => {
    await stylelint.utils.checkAgainstRule(
      {
        ruleName: 'media-feature-name-allowed-list',
        ruleSettings,
        root,
      },
      (warning) => {
        stylelint.utils.report({
          ruleName,
          result,
          node: warning.node,
          message: warning.text,
        });
      },
    );
  };
);
```

### What minimal configuration is needed to reproduce the bug?

```js
export default {
  plugins: ['./rules/custom-rule-one', './rules/custom-rule-two'],
  rules: { 'custom/rule-one': settings },
}
```

### How did you run Stylelint?

```shell
stylelint "**/*.css"
```

### Which Stylelint-related dependencies are you using?

```json
{
  "devDependencies": {
    "stylelint": "^16.6.1"
  }
}
```

### What did you expect to happen?

Expected `checkAgainstRule` to report warnings from resolved async rules

### What actually happened?

No errors were reported

### Do you have a proposal to fix the bug?

Update [`checkAgainstRule.mjs`](https://github.com/stylelint/stylelint/blob/cb3a7f8d5e89f2e467a4166293c17d3c2d1e72cc/lib/utils/checkAgainstRule.mjs#L58) and [`checkAgainstRule.cjs`](https://github.com/stylelint/stylelint/blob/cb3a7f8d5e89f2e467a4166293c17d3c2d1e72cc/lib/utils/checkAgainstRule.cjs#L61) to await the `ruleFunc` call.

**Comment by ybiquitous** [comment]
@aaronccasanova Thanks for the report. I can reproduce the bug with the following code:

`custom-rule-one.mjs`:

```js
import stylelint from 'stylelint';

const { createPlugin, utils } = stylelint;

const ruleName = 'custom/rule-one';

const rule = (ruleSettings) => async (root, result) => {
	await utils.checkAgainstRule(
		{
			ruleName: 'custom/rule-two',
			ruleSettings,
			root,
			result,
		},
		(warning) => {
			utils.report({
				ruleName,
				result,
				node: warning.node,
				message: warning.text,
			});
		},
	);
};
rule.ruleName = ruleName;

export default createPlugin(ruleName, rule);
```

`custom-rule-two.mjs`:

```js
import stylelint from 'stylelint';

const { createPlugin, utils } = stylelint;

const ruleName = 'custom/rule-two';

const rule = (ruleSettings) => async (root, result) => {
	await utils.checkAgainstRule(
		{
			ruleName: 'media-query-no-invalid',
			ruleSettings,
			root,
			result,
		},
		(warning) => {
			utils.report({
				ruleName,
				result,
				node: warning.node,
				message: warning.text,
			});
		},
	);
};
rule.ruleName = ruleName;

export default createPlugin(ruleName, rule);
```

`stylelint.config.mjs`:

```js
import customRuleOne from './custom-rule-one.mjs';
import customRuleTwo from './custom-rule-two.mjs';

export default {
	plugins: [customRuleOne, customRuleTwo],
	rules: {
		'custom/rule-one': true,
	},
};
```

Run:

```sh-session
$ echo '@media invalid() {}' | npx stylelint -f verbose

1 source checked
  <input css pjCFij>

0 problems found
```
